### ISSUE TITLE
French model sentences segmentation problem

### ISSUE THREAD

**Issue by cglacet** [explosion/spaCy#4637]
## The problem 

The character "-" is recognized as a sentence separator which is odd since it's never used as a sentence separator in French.

## How to reproduce the behaviour

```python
>>> import spacy
>>> nlp = spacy.load('fr_core_news_sm')
>>> doc = nlp("Bon, me dis-je, s'il a faim, mon oncle, qui est le plus impatient des hommes, va pousser des cris de détresse.")
>>> next(doc.sents)
`Bon, me dis`
```

## Your Environment

* **spaCy version:** 2.2.2
* **Platform:** Darwin-18.6.0-x86_64-i386-64bit
* **Python version:** 3.7.3

**Comment by adrianeboyd** [comment]
The sentence boundaries are predicted by the parser and statistical models like the parser model are not going to be 100% correct. (See #3052, which is the best place to mention issues like this!)

I checked out of curiosity and our training data (from an older release of https://github.com/UniversalDependencies/UD_French-Sequoia) does include quite a few sentences starting with `-`, which look like they're originally from things like lists in Wikipedia articles. By a very rough count, 75% of `-` tokens start a sentence in the data, which may be unusual, but that explains why the parsing model has learned to start sentences with `-`.

**Comment by cglacet** [comment]
That's very strange, I'm curious too and I wonder in which cases a **-** (small dash) will be used to end a phrase. I think there is a problem in the corpus if that's the case. In french, this short **-** is only used in two cases: for a word composed of several words ([*mot-valise*](https://www.wikiwand.com/fr/Mot-valise)/[*clitique*](https://www.wikiwand.com/fr/Clitique)) or for words that have been split in two because they are too long to appear on a single line. In both these cases, **-** doesn't end the sentence. 

On the other hand, **–** (medium dash) can be used in lists or, in very rare occasions, it can be used to replace parenthesis. 

Finally **—** (long dash) is mostly used in dialogues or, once again, to replace parenthesis (this is a bit more common than **–**). 

It's very likely that the corpus contains some texts where **-** are misused but 75% looks like there is something else hiding. 

Even in cases where **—** or **–** should be used, **-** wouldn't mark the end of a sentence (nor the beginning). For example with the following dialogue (from this [very broken sample](https://www.gutenberg.org/files/18143/18143-h/18143-h.htm)): 

> JULIETTE.—Roméo!
> ROMÉO.—Ma douce amie!
> JULIETTE.—A quelle heure demain matin enverrai-je vers toi?
> ROMÉO.—A neuf heures.
>JULIETTE.—Je n'y manquerai pas: d'ici à ce moment il y a vingt années..... J'ai oublié pourquoi je t'ai rappelé.

**—** doesn't end the sentence « Ma douce amie! » and it doesn't start the sentence « A quelle heure demain matin enverrai-je vers toi? ». Both sentences are split in the most classical way, they start with an uppercase letter and end with a punctuation sign `!`. The **—** is only here to highlight that this block of text is part of a dialogue or that the speaker changes. 

The structure is very different from what we saw in «Bon, me dis-je, s'il […] », where there is no uppercase, no line break, no punctuation. I don't know how your model is trained, but I'm pretty sure characters are not isolated from their context and I think this shouldn't happen. 

The best criteria to find a sentence is by far the presence of punctuation and uppercase letter. It's so omnipresent that your model will use this for sure. The only hard cases I can think of for your training are « Bonjour Jean-Sébastien, comment allez-vous ? » from « Bonjour Jean ! Sébastien, comment allez-vous ? ». In these cases, only a very good knowledge of context could help. I wouldn't have been surprised if splitting had failed on such a case.

In the end I guess that training a model for predicting sentences is just an exercise and that a hard coded solution would have better results (using only grammar, typography and context to programmatically split a text).

**Comment by adrianeboyd** [comment]
I think the answer is simply that a list bullet point has been normalized to `-` in this treebank. It doesn't mark the end of a sentence, it's used at the beginning of a sentence. Since `-` isn't used in many other contexts, it's a strong clue for the parser to learn that this is the start of a sentence. This may be atypical, but it's the data that we have available with a license that allows us to distribute pretrained models that can be used for commercial purposes.

You are welcome to train a parser with a different corpus or to develop a rule-based sentence boundary component for French. One example of a rule-based SBD tool for English is pySBD, which also shows how a component like this can be set up for easy integration with spacy. I'm not very familiar with French NLP tools, but there may already be tools for French like pySBD that are easy to use with spacy, too.

Since the choices within spacy for sentence segmentation are somewhat limited (fast but overly simple sentencizer or slow parser), I'm starting work on a new statistical sentence segmenter that just learns sentence boundaries, which will make it easier to train models with a wider range of data, and it will also hopefully be faster than the parser. We'll see if it can improve the performance for French in cases like this...
